src/main/java/org/apache/commons/lang3/ArrayUtils.java

```diff
@@ -1432,6 +1432,115 @@ public static <T> T arraycopy(final T source, final int sourcePos, final T dest,
         return dest;
     }
 
+    /**
+     * Searches element in array sorted by key.
+     *
+     * @param array
+     *      array sorted by key field
+     * @param key
+     *      key to search for
+     * @param keyExtractor
+     *      function to extract key from element
+     * @param comparator
+     *      comparator for keys
+     *
+     * @return
+     *      index of the search key, if it is contained in the array within specified range; otherwise,
+     *      (-first_greater - 1).  The first_greater is the index of lowest greater element in the list - if all elements
+     *      are lower, the first_greater is defined as toIndex.
```

Comment:
The result should be undefined if all array elements are equals; like `Arrays` version.

Reply:
Updated to match non-index version.

@garydgregory : That is great point.  In fact, one of my use cases is array (list) with duplicities.  So this can be done better.  I introduce `binarySearchFirst` and `binarySearchLast` methods which return consistently first and last matching element respectively.

Reply:
WRT binarySearchFirst and Last, let's not add those.

Reply:
@garydgregory : I already did.  I dropped `binarySearch` as the randomness of choosing any of equal element effectively disqualifies it from such use cases.  Predictable behavior of `binarySearchFirst` and `binarySearchLast` is huge benefit and comes at almost no cost (is slower only in cases when the lookup is quickly "lucky").

Reply:
The use caee, I thought, was a version of the 'Arrays' method plus key extraction. I would not necessarily want an API that always allocates an array to return a value. Maybe do this later, as a second API only if needed.

Reply:
> Sudden thought: surely key extraction can take place in the Comparator, so why is this method needed at all?
> 
> Or maybe I am missing something?

Good point @sebbASF !

We might not need this at all...

Reply:
@sebbASF @garydgregory :

> In the case of the 'Arrays' method, the entire entry is matched, so to some extent it does not matter which index is returned. That is not the case where a partial key is used, so I think in the user will often want to know all the matches.

In my particular case, I have list of GPS locations sorted by time and track distance, with distance unchanged if there was no move (causing duplicates).  One use case is to find point for particular distance and if there are duplicates, I want the result to be predictable (here last but can be first in different use cases).  If the result is unpredictable, it will lead to different results based on initial conditions for the search.  I don't care about whole range, though I can imagine someone would appreciate such function with added complexity.

> Sudden thought: surely key extraction can take place in the Comparator, so why is this method needed at all?

No, unless we're able to construct the full object which is not always possible - note that `Comparator` takes two arguments of the same type.  It is technically achievable with hacking and casting one argument to `T` and the other one to `K`, relying on internal implementation passing the arguments in expected order.

@garydgregory :
> Please revert the First/Last APIs.

For the reasons above, I want to keep those as they better match intended use case.  I think we actually don't need original unpredictable `binarySearch()` .